File: webapp/src/lib/kofiaFreeSis.ts
/**
 * KOFIA FreeSIS (금융투자협회 종합통계) — market fund & forced-sell series.
 * 증시자금추이(STATSCU0100000060): 예탁금·미수금·반대매매.
 * 신용공여 잔고(STATSCU0100000070): 신용거래융자·대주.
 */

const UA =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36";

const FREE_SIS = "https://freesis.kofia.or.kr";
const FUND_SERVICE = "STATSCU0100000060";
const CREDIT_SERVICE = "STATSCU0100000070";
/** FreeSIS 단위 콤보 코드 — '01'일 때 원 스케일이 네이버 증시자금(억원)과 일치 */
const UNIT_WON_SCALE = "01";

export type KofiaFundDay = {
  date: string; // YYYY-MM-DD
  deposit: number; // 원 — 투자자예탁금(파생예수금 제외)
  deriv_deposit: number;
  rp_balance: number;
  unsettled: number; // 위탁매매 미수금
  opp_sell: number; // 실제 반대매매금액(미수 기준)
  opp_ratio_pct: number; // 미수금 대비 반대매매비중(%)
};

export type KofiaCreditDay = {
  date: string;
  /** 신용거래융자 전체 (원) */
  loan_total: number;
  loan_kospi: number;
  loan_kosdaq: number;
  /** 신용거래대주 전체 (원) */
  short_total: number;
  collateral_loan: number;
};

export type ForcedSellStress = "calm" | "elevated" | "high" | "extreme";

export type ForcedSellBoard = {
  as_of: string | null;
  stress: ForcedSellStress;
  stress_label: string;
  latest_fund: KofiaFundDay | null;
  latest_credit: KofiaCreditDay | null;
  /** Day-over-day credit loan change (원); negative ≈ 신용 축소 */
  credit_delta: number | null;
  fund_series: KofiaFundDay[];
  credit_series: KofiaCreditDay[];
  note: string;
};

function ymdToIso(ymd: string): string {
  const s = String(ymd).replace(/\D/g, "");
  if (s.length !== 8) return String(ymd);
  return `${s.slice(0, 4)}-${s.slice(4, 6)}-${s.slice(6, 8)}`;
}

function num(v: unknown): number {
  if (typeof v === "number" && Number.isFinite(v)) return v;
  if (v == null) return 0;
  const n = Number(String(v).replace(/,/g, ""));
  return Number.isFinite(n) ? n : 0;
}

function stressFromRatio(ratio: number): { stress: ForcedSellStress; label: string } {
  if (ratio >= 10) return { stress: "extreme", label: "강제매도 압력 매우 큼" };
  if (ratio >= 5) return { stress: "high", label: "강제매도 압력 큼" };
  if (ratio >= 2) return { stress: "elevated", label: "다소 높음" };
  return { stress: "calm", label: "안정" };
}

function cookieFrom(res: Response, prev = ""): string {
  const map = Object.fromEntries(
    (prev ? prev.split("; ") : [])
      .filter(Boolean)
      .map((x) => {
        const i = x.indexOf("=");
        return [x.slice(0, i), x.slice(i + 1)] as const;
      }),
  );
  const raw: string[] = [];
  if (typeof res.headers.getSetCookie === "function") {
    raw.push(...res.headers.getSetCookie());
  }
  const single = res.headers.get("set-cookie");
  if (single) raw.push(single);
  for (const c of raw) {
    const nv = c.split(";")[0];
    const i = nv.indexOf("=");
    if (i > 0) map[nv.slice(0, i)] = nv.slice(i + 1);
  }
  return Object.entries(map)
    .map(([k, v]) => `${k}=${v}`)
    .join("; ");
}

/** FreeSIS/Java sometimes emits NaN/Infinity which Strict JSON.parse rejects. */
function parseFreesisJson<T>(text: string): T {
  const cleaned = text
    .replace(/:\s*NaN\b/g, ":null")
    .replace(/:\s*-Infinity\b/g, ":null")
    .replace(/:\s*Infinity\b/g, ":null")
    .replace(/:\s*undefined\b/g, ":null");
  return JSON.parse(cleaned) as T;
}

async function freesisPost<T>(
  path: string,
  body: unknown,
  cookie: string,
): Promise<{ json: T; cookie: string }> {
  const res = await fetch(`${FREE_SIS}${path}`, {
    method: "POST",
    headers: {
      "User-Agent": UA,
      "Content-Type": "application/json",
      Accept: "application/json",
      "X-Requested-With": "XMLHttpRequest",
      Referer: `${FREE_SIS}/stat/FreeSIS.do?parentDivId=MSIS10000000000000&serviceId=${FUND_SERVICE}`,
      Origin: FREE_SIS,
      Cookie: cookie,
    },
    body: JSON.stringify(body),
    cache: "no-store",
  });
  const next = cookieFrom(res, cookie);
  const ct = res.headers.get("content-type") || "";
  const text = await res.text();
  const looksJson =
    ct.includes("json") ||
    text.trimStart().startsWith("{") ||
    text.trimStart().startsWith("[");
  if (!looksJson) {
    throw new Error(`FreeSIS ${path} non-JSON (${res.status})`);
  }
  try {
    return { json: parseFreesisJson<T>(text), cookie: next };
  } catch (exc) {
    throw new Error(
      `FreeSIS ${path} JSON parse failed: ${exc instanceof Error ? exc.message : "unknown"}`,
    );
  }
}

function kstYmd(ms: number): string {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: "Asia/Seoul",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  })
    .format(new Date(ms))
    .replace(/-/g, "");
}

function lookbackRange(days: number): { start: string; end: string } {
  const endMs = Date.now();
  return {
    start: kstYmd(endMs - days * 86_400_000),
    end: kstYmd(endMs),
  };
}

type MetaList = {
  ds1?: Array<Record<string, unknown>>;
};

type MetaSrv = {
  dsGrid?: Array<{ ORDER_SEQ?: number; HEADER_ID?: string; HEADER_NM?: string }>;
  dsGridServlet?: Array<{ OBJ_NM?: string }>;
};

async function warmSession(): Promise<string> {
  const res = await fetch(
    `${FREE_SIS}/stat/FreeSIS.do?parentDivId=MSIS10000000000000&serviceId=${FUND_SERVICE}`,
    {
      headers: { "User-Agent": UA, Accept: "text/html" },
      cache: "no-store",
    },
  );
  return cookieFrom(res);
}

function parseFundRows(rows: Array<Record<string, unknown>>): KofiaFundDay[] {
  const out: KofiaFundDay[] = [];
  for (const r of rows) {
    const date = ymdToIso(String(r.TMPV1 ?? ""));
    if (!/^\d{4}-\d{2}-\d{2}$/.test(date)) continue;
    const opp_ratio_pct = num(r.TMPV7);
    const deposit = num(r.TMPV2);
    const unsettled = num(r.TMPV5);
    const opp_sell = num(r.TMPV6);
    // Need at least ratio or an amount field to keep the row.
    if (!deposit && !unsettled && !opp_sell && !opp_ratio_pct) continue;
    out.push({
      date,
      deposit,
      deriv_deposit: num(r.TMPV3),
      rp_balance: num(r.TMPV4),
      unsettled,
      opp_sell,
      opp_ratio_pct,
    });
  }
  return out.sort((a, b) => a.date.localeCompare(b.date));
}

function mapCreditRows(
  rows: Array<Record<string, unknown>>,
  _headers: Array<{ ORDER_SEQ?: number; HEADER_NM?: string }>,
): KofiaCreditDay[] {
  // STATSCU0100000070 fixed layout (group headers excluded from TMPV*):
  // 1 구분 · 2 융자전체 · 3 융자코스피 · 4 융자코스닥 ·
  // 5 대주전체 · 6 대주코스피 · 7 대주코스닥 · 8 청약 · 9 예탁담보
  void _headers;
  const out: KofiaCreditDay[] = [];
  for (const r of rows) {
    const date = ymdToIso(String(r.TMPV1 ?? ""));
    if (!/^\d{4}-\d{2}-\d{2}$/.test(date)) continue;
    const loan_total = num(r.TMPV2);
    if (!loan_total) continue;
    out.push({
      date,
      loan_total,
      loan_kospi: num(r.TMPV3),
      loan_kosdaq: num(r.TMPV4),
      short_total: num(r.TMPV5),
      collateral_loan: num(r.TMPV9),
    });
  }
  return out.sort((a, b) => a.date.localeCompare(b.date));
}

async function fetchFundSeries(
  cookie: string,
  start: string,
  end: string,
): Promise<{ rows: KofiaFundDay[]; cookie: string }> {
  // Meta warm (browser does this before grid submit)
  const meta = await freesisPost<MetaSrv>(
    "/meta/getSrvData.do",
    {
      dmSearchData: {
        strSvrId: FUND_SERVICE,
        tmpV1: "RD",
        tmpV45: start,
        tmpV46: end,
        strGetCode: "Y",
      },
    },
    cookie,
  );
  const obj = meta.json.dsGridServlet?.[0]?.OBJ_NM || `${FUND_SERVICE}BO`;
  const list = await freesisPost<MetaList>(
    "/meta/getMetaDataList.do",
    {
      dmSearch: {
        strSvrId: FUND_SERVICE,
        tmpV1: "RD",
        tmpV40: UNIT_WON_SCALE,
        tmpV45: start,
        tmpV46: end,
        OBJ_NM: obj,
      },
    },
    meta.cookie,
  );
  return { rows: parseFundRows(list.json.ds1 || []), cookie: list.cookie };
}

async function fetchCreditSeries(
  cookie: string,
  start: string,
  end: string,
): Promise<{ rows: KofiaCreditDay[]; cookie: string }> {
  const meta = await freesisPost<MetaSrv>(
    "/meta/getSrvData.do",
    {
      dmSearchData: {
        strSvrId: CREDIT_SERVICE,
        tmpV1: "RD",
        tmpV45: start,
        tmpV46: end,
        strGetCode: "Y",
      },
    },
    cookie,
  );
  const obj = meta.json.dsGridServlet?.[0]?.OBJ_NM || `${CREDIT_SERVICE}BO`;
  const headers = meta.json.dsGrid || [];
  const list = await freesisPost<MetaList>(
    "/meta/getMetaDataList.do",
    {
      dmSearch: {
        strSvrId: CREDIT_SERVICE,
        tmpV1: "RD",
        tmpV40: UNIT_WON_SCALE,
        tmpV45: start,
        tmpV46: end,
        OBJ_NM: obj,
      },
    },
    meta.cookie,
  );
  return {
    rows: mapCreditRows(list.json.ds1 || [], headers),
    cookie: list.cookie,
  };
}

export async function fetchForcedSellBoard(
  lookbackDays = 60,
): Promise<ForcedSellBoard> {
  const emptyNote =
    "금융투자협회 FreeSIS 증시자금·신용공여. 미수 기준 반대매매(신용담보 반대매매 전체와 범위가 다를 수 있음). 통상 1~2영업일 지연.";

  const empty = (extra?: string): ForcedSellBoard => ({
    as_of: null,
    stress: "calm",
    stress_label: "데이터 없음",
    latest_fund: null,
    latest_credit: null,
    credit_delta: null,
    fund_series: [],
    credit_series: [],
    note: extra ? `${emptyNote} ${extra}` : emptyNote,
  });

  const attempt = async (): Promise<ForcedSellBoard> => {
    const { start, end } = lookbackRange(lookbackDays);
    if (!/^\d{8}$/.test(start) || !/^\d{8}$/.test(end)) {
      throw new Error("invalid lookback dates");
    }
    let cookie = await warmSession();
    const fund = await fetchFundSeries(cookie, start, end);
    cookie = fund.cookie;
    let creditRows: KofiaCreditDay[] = [];
    try {
      const credit = await fetchCreditSeries(cookie, start, end);
      creditRows = credit.rows;
    } catch {
      creditRows = [];
    }

    const fund_series = fund.rows;
    const credit_series = creditRows;
    if (!fund_series.length && !credit_series.length) {
      throw new Error("empty FreeSIS series");
    }
    const latest_fund = fund_series[fund_series.length - 1] || null;
    const latest_credit = credit_series[credit_series.length - 1] || null;
    const prevCredit = credit_series[credit_series.length - 2];
    const credit_delta =
      latest_credit && prevCredit
        ? latest_credit.loan_total - prevCredit.loan_total
        : null;
    const { stress, label } = stressFromRatio(latest_fund?.opp_ratio_pct ?? 0);

    return {
      as_of: latest_fund?.date || latest_credit?.date || null,
      stress,
      stress_label: label,
      latest_fund,
      latest_credit,
      credit_delta,
      fund_series,
      credit_series,
      note: emptyNote,
    };
  };

  try {
    return await attempt();
  } catch {
    try {
      // One retry — FreeSIS sessions / NaN payloads are intermittently flaky.
      return await attempt();
    } catch {
      return empty("지금은 협회 통계를 불러오지 못했습니다. 잠시 후 새로고침해 주세요.");
    }
  }
}

export function fmtWonEok(n?: number | null, digits = 0): string {
  if (n == null || !Number.isFinite(n)) return "—";
  const eok = n / 1e8;
  return `${eok.toLocaleString("ko-KR", {
    maximumFractionDigits: digits,
    minimumFractionDigits: digits,
  })}억`;
}

export function fmtWonJo(n?: number | null): string {
  if (n == null || !Number.isFinite(n)) return "—";
  const jo = n / 1e12;
  return `${jo.toLocaleString("ko-KR", { maximumFractionDigits: 1 })}조`;
}

File: kofia_freesis.py
"""KOFIA FreeSIS + Naver 증시자금 scraper (Render egress).

Ports webapp/src/lib/kofiaFreeSis.ts so 반대매매 can be collected outside
Vercel US IP ranges, then mirrored to R2 via r2_data.upload_credit_monitor.
"""

from __future__ import annotations

import json
import re
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any
from zoneinfo import ZoneInfo

import requests

KST = ZoneInfo("Asia/Seoul")
UA = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
)
FREE_SIS = "https://freesis.kofia.or.kr"
FUND_SERVICE = "STATSCU0100000060"
CREDIT_SERVICE = "STATSCU0100000070"
UNIT_WON_SCALE = "01"
TIMEOUT = 20

DATA_DIR = Path(__file__).resolve().parent / "data" / "credit_monitor"
LATEST_PATH = DATA_DIR / "latest.json"


def _num(v: Any) -> float:
    if isinstance(v, (int, float)) and v == v:  # not NaN
        return float(v)
    if v is None:
        return 0.0
    try:
        return float(str(v).replace(",", ""))
    except (TypeError, ValueError):
        return 0.0


def _ymd_to_iso(ymd: str) -> str:
    s = re.sub(r"\D", "", str(ymd))
    if len(s) != 8:
        return str(ymd)
    return f"{s[:4]}-{s[4:6]}-{s[6:8]}"


def _lookback_range(days: int) -> tuple[str, str]:
    end = datetime.now(KST).date()
    start = end - timedelta(days=days)
    return start.strftime("%Y%m%d"), end.strftime("%Y%m%d")


def _stress_from_ratio(ratio: float) -> tuple[str, str]:
    if ratio >= 10:
        return "extreme", "강제매도 압력 매우 큼"
    if ratio >= 5:
        return "high", "강제매도 압력 큼"
    if ratio >= 2:
        return "elevated", "다소 높음"
    return "calm", "안정"


def _parse_freesis_json(text: str) -> Any:
    cleaned = (
        text.replace(": NaN", ": null")
        .replace(":NaN", ":null")
        .replace(": -Infinity", ": null")
        .replace(":-Infinity", ":null")
        .replace(": Infinity", ": null")
        .replace(":Infinity", ":null")
        .replace(": undefined", ": null")
        .replace(":undefined", ":null")
    )
    # also handle word-boundary NaN without space variants already covered
    cleaned = re.sub(r":\s*NaN\b", ":null", cleaned)
    cleaned = re.sub(r":\s*-?Infinity\b", ":null", cleaned)
    return json.loads(cleaned)


class FreeSisClient:
    def __init__(self) -> None:
        self.session = requests.Session()
        self.session.headers.update(
            {
                "User-Agent": UA,
                "Accept": "application/json, text/html, */*",
                "X-Requested-With": "XMLHttpRequest",
                "Origin": FREE_SIS,
            }
        )

    def warm(self) -> None:
        url = (
            f"{FREE_SIS}/stat/FreeSIS.do"
            f"?parentDivId=MSIS10000000000000&serviceId={FUND_SERVICE}"
        )
        self.session.get(url, timeout=TIMEOUT)

    def post(self, path: str, body: dict[str, Any]) -> Any:
        res = self.session.post(
            f"{FREE_SIS}{path}",
            json=body,
            headers={
                "Content-Type": "application/json",
                "Referer": (
                    f"{FREE_SIS}/stat/FreeSIS.do"
                    f"?parentDivId=MSIS10000000000000&serviceId={FUND_SERVICE}"
                ),
            },
            timeout=TIMEOUT,
        )
        text = res.text
        if not (
            "json" in (res.headers.get("content-type") or "").lower()
            or text.lstrip().startswith("{")
            or text.lstrip().startswith("[")
        ):
            raise RuntimeError(f"FreeSIS {path} non-JSON ({res.status_code})")
        try:
            return _parse_freesis_json(text)
        except json.JSONDecodeError as exc:
            raise RuntimeError(f"FreeSIS {path} JSON parse failed: {exc}") from exc


def _parse_fund_rows(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for r in rows:
        date = _ymd_to_iso(str(r.get("TMPV1") or ""))
        if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", date):
            continue
        opp_ratio_pct = _num(r.get("TMPV7"))
        deposit = _num(r.get("TMPV2"))
        unsettled = _num(r.get("TMPV5"))
        opp_sell = _num(r.get("TMPV6"))
        if not deposit and not unsettled and not opp_sell and not opp_ratio_pct:
            continue
        out.append(
            {
                "date": date,
                "deposit": deposit,
                "deriv_deposit": _num(r.get("TMPV3")),
                "rp_balance": _num(r.get("TMPV4")),
                "unsettled": unsettled,
                "opp_sell": opp_sell,
                "opp_ratio_pct": opp_ratio_pct,
            }
        )
    out.sort(key=lambda x: x["date"])
    return out


def _parse_credit_rows(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for r in rows:
        date = _ymd_to_iso(str(r.get("TMPV1") or ""))
        if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", date):
            continue
        loan_total = _num(r.get("TMPV2"))
        if not loan_total:
            continue
        out.append(
            {
                "date": date,
                "loan_total": loan_total,
                "loan_kospi": _num(r.get("TMPV3")),
                "loan_kosdaq": _num(r.get("TMPV4")),
                "short_total": _num(r.get("TMPV5")),
                "collateral_loan": _num(r.get("TMPV9")),
            }
        )
    out.sort(key=lambda x: x["date"])
    return out


def fetch_fund_series(client: FreeSisClient, start: str, end: str) -> list[dict[str, Any]]:
    meta = client.post(
        "/meta/getSrvData.do",
        {
            "dmSearchData": {
                "strSvrId": FUND_SERVICE,
                "tmpV1": "RD",
                "tmpV45": start,
                "tmpV46": end,
                "strGetCode": "Y",
            }
        },
    )
    servlet = (meta.get("dsGridServlet") or [{}])[0]
    obj = servlet.get("OBJ_NM") or f"{FUND_SERVICE}BO"
    listing = client.post(
        "/meta/getMetaDataList.do",
        {
            "dmSearch": {
                "strSvrId": FUND_SERVICE,
                "tmpV1": "RD",
                "tmpV40": UNIT_WON_SCALE,
                "tmpV45": start,
                "tmpV46": end,
                "OBJ_NM": obj,
            }
        },
    )
    return _parse_fund_rows(list(listing.get("ds1") or []))


def fetch_credit_series(client: FreeSisClient, start: str, end: str) -> list[dict[str, Any]]:
    meta = client.post(
        "/meta/getSrvData.do",
        {
            "dmSearchData": {
                "strSvrId": CREDIT_SERVICE,
                "tmpV1": "RD",
                "tmpV45": start,
                "tmpV46": end,
                "strGetCode": "Y",
            }
        },
    )
    servlet = (meta.get("dsGridServlet") or [{}])[0]
    obj = servlet.get("OBJ_NM") or f"{CREDIT_SERVICE}BO"
    listing = client.post(
        "/meta/getMetaDataList.do",
        {
            "dmSearch": {
                "strSvrId": CREDIT_SERVICE,
                "tmpV1": "RD",
                "tmpV40": UNIT_WON_SCALE,
                "tmpV45": start,
                "tmpV46": end,
                "OBJ_NM": obj,
            }
        },
    )
    return _parse_credit_rows(list(listing.get("ds1") or []))


def fetch_naver_credit_series() -> tuple[list[dict[str, float]], list[dict[str, Any]]]:
    res = requests.get(
        "https://finance.naver.com/sise/sise_deposit.naver",
        headers={
            "User-Agent": UA,
            "Accept": "text/html,application/xhtml+xml",
            "Referer": "https://finance.naver.com/",
            "Accept-Language": "ko-KR,ko;q=0.9",
        },
        timeout=TIMEOUT,
    )
    res.raise_for_status()
    # Naver often serves EUC-KR
    raw = res.content
    try:
        html = raw.decode("euc-kr")
    except UnicodeDecodeError:
        html = raw.decode("utf-8", errors="replace")

    deposit_series: list[dict[str, float]] = []
    credit_series: list[dict[str, Any]] = []
    for tr in re.findall(r"<tr[^>]*>[\s\S]*?</tr>", html, flags=re.I):
        cells = [
            re.sub(r"\s+", " ", re.sub(r"<[^>]+>", " ", c)).strip()
            for c in re.findall(r"<t[dh][^>]*>[\s\S]*?</t[dh]>", tr, flags=re.I)
        ]
        if not cells or not re.fullmatch(r"\d{2}\.\d{2}\.\d{2}", cells[0]):
            continue
        date = "20" + cells[0].replace(".", "-")
        deposit_eok = _num(cells[1] if len(cells) > 1 else 0)
        credit_eok = _num(cells[3] if len(cells) > 3 else 0)
        if not deposit_eok and not credit_eok:
            continue
        deposit = deposit_eok * 1e8
        loan_total = credit_eok * 1e8
        deposit_series.append({"date": date, "deposit": deposit})
        credit_series.append(
            {
                "date": date,
                "loan_total": loan_total,
                "loan_kospi": 0,
                "loan_kosdaq": 0,
                "short_total": 0,
                "collateral_loan": 0,
            }
        )
    deposit_series.sort(key=lambda x: x["date"])
    credit_series.sort(key=lambda x: x["date"])
    return deposit_series, credit_series


def try_freesis(start: str, end: str) -> tuple[list[dict[str, Any]], list[dict[str, Any]], str | None]:
    try:
        client = FreeSisClient()
        client.warm()
        fund = fetch_fund_series(client, start, end)
        credit: list[dict[str, Any]] = []
        try:
            credit = fetch_credit_series(client, start, end)
        except Exception:
            credit = []
        return fund, credit, None
    except Exception as exc:
        return [], [], str(exc)


def build_forced_sell_board(lookback_days: int = 60) -> dict[str, Any]:
    base_note = (
        "미수 기준 반대매매는 금투협 FreeSIS, 신용·예탁금은 FreeSIS 또는 네이버 증시자금. "
        "통상 1~2영업일 지연. Render→R2 캐시."
    )
    start, end = _lookback_range(lookback_days)
    fund, freesis_credit, freesis_error = try_freesis(start, end)

    deposit_series: list[dict[str, float]] = []
    naver_credit: list[dict[str, Any]] = []
    try:
        deposit_series, naver_credit = fetch_naver_credit_series()
    except Exception:
        pass

    credit_series = freesis_credit if freesis_credit else naver_credit
    fund_for_ui = fund
    if not fund_for_ui and deposit_series:
        fund_for_ui = [
            {
                "date": d["date"],
                "deposit": d["deposit"],
                "deriv_deposit": 0,
                "rp_balance": 0,
                "unsettled": 0,
                "opp_sell": 0,
                "opp_ratio_pct": 0,
            }
            for d in deposit_series
        ]

    sources = {
        "freesis_fund": len(fund) > 0,
        "freesis_credit": len(freesis_credit) > 0,
        "naver_credit": len(freesis_credit) == 0 and len(naver_credit) > 0,
        "r2": False,
    }

    generated_at = datetime.now(KST).isoformat()

    if not fund_for_ui and not credit_series:
        return {
            "as_of": None,
            "stress": "calm",
            "stress_label": "데이터 없음",
            "latest_fund": None,
            "latest_credit": None,
            "credit_delta": None,
            "fund_series": [],
            "credit_series": [],
            "sources": sources,
            "note": f"{base_note} 지금은 통계를 불러오지 못했습니다."
            + (f" ({freesis_error})" if freesis_error else ""),
            "generated_at": generated_at,
            "collected_on": "render",
        }

    latest_fund_kpi = fund[-1] if sources["freesis_fund"] else None
    latest_fund = latest_fund_kpi or (fund_for_ui[-1] if fund_for_ui else None)
    latest_credit = credit_series[-1] if credit_series else None
    prev_credit = credit_series[-2] if len(credit_series) >= 2 else None
    credit_delta = (
        (latest_credit["loan_total"] - prev_credit["loan_total"])
        if latest_credit and prev_credit
        else None
    )

    if sources["freesis_fund"]:
        stress, label = _stress_from_ratio(float((latest_fund_kpi or {}).get("opp_ratio_pct") or 0))
    else:
        stress, label = "calm", "신용만 표시"

    parts = [base_note]
    if sources["freesis_fund"]:
        parts.append("반대매매: FreeSIS")
    else:
        parts.append("반대매매: FreeSIS 접속 불가")
    if sources["freesis_credit"]:
        parts.append("신용: FreeSIS")
    elif sources["naver_credit"]:
        parts.append("신용·예탁금: 네이버 증시자금")

    return {
        "as_of": (latest_fund_kpi or latest_credit or latest_fund or {}).get("date")
        if (latest_fund_kpi or latest_credit or latest_fund)
        else None,
        "stress": stress,
        "stress_label": label,
        "latest_fund": latest_fund_kpi,
        "latest_credit": latest_credit,
        "credit_delta": credit_delta,
        "fund_series": fund if sources["freesis_fund"] else [],
        "credit_series": credit_series,
        "sources": sources,
        "note": " · ".join(parts),
        "generated_at": generated_at,
        "collected_on": "render",
    }


def save_local(board: dict[str, Any]) -> Path:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    LATEST_PATH.write_text(json.dumps(board, ensure_ascii=False), encoding="utf-8")
    return LATEST_PATH


def collect_and_publish(*, lookback_days: int = 60) -> dict[str, Any]:
    board = build_forced_sell_board(lookback_days=lookback_days)
    save_local(board)
    published = False
    pub_error = None
    try:
        from r2_data import upload_credit_monitor

        published = upload_credit_monitor(board)
    except Exception as exc:
        pub_error = str(exc)
        print(f"credit_monitor R2 upload failed: {exc}")
    return {
        "board": board,
        "r2": published,
        "error": pub_error,
        "as_of": board.get("as_of"),
        "freesis_fund": bool((board.get("sources") or {}).get("freesis_fund")),
    }
